### ISSUE TITLE
[Feature] Transparent Wrapper Components in Aurelia

### ISSUE THREAD

**Issue by HamedFathi** [aurelia/aurelia#319]
# 🙋 Feature Request

One of the most exciting features in `Vue.js` is called **Transparent Wrapper Components**.
You can find some information from [here](https://vuejs.org/v2/guide/components-custom-events.html#Binding-Native-Events-to-Components).

When we were developing Aurelia-Toolbelt, we were faced with this problem a lot. For example, a new component has actually developed a base component such as **input**

I wrote [Aurelia-Toolbelt Float-Input](https://aurelia-toolbelt.github.io/#/bootstrap/float-input)

You can use this like

```html
<abt-float-input type='email' placeholder='email@example.com'>Email</abt-float-input>
```

Suppose I want to add maxlength="10" to this element.

```html
<abt-float-input type='email'  maxlength="10" placeholder='email@example.com'>Email</abt-float-input>
```

As you know `abt-float-input` is nothing more than an extended input element. so I want to transfer all dedicated attributes of an input HTML element to input element of the `float-input` template.

![image](https://user-images.githubusercontent.com/8418700/49019588-7f269b80-f1a3-11e8-89b8-9c54f1f2d06b.png)

I think this is very very very useful for creating any UI component. We can rewrite all Aurelia-Toolbelt components with the benefit of HTML existing attributes.

## 🤔 Expected Behavior

I can use any dedicated attributes of HTML besides my new features/bindings/attributes.

## 😯 Current Behavior

I should implement all attributes one by one. 

## 💁 Possible Solution

I have not enough knowledge about how to add this feature. but @bigopon helps me about this in [here](https://discourse.aurelia.io/t/assign-all-new-attributes-to-specific-element-transparent-wrapper-components)

We want to create custom attributes eg `au-bind`. We can assign `au-bind` to target element inside a template when developing an element. after that, It should bind all attributes of a new custom element to the target element.

![image](https://user-images.githubusercontent.com/8418700/49020360-7cc54100-f1a5-11e8-8887-b9a7f24b15e5.png)

## 💻 Examples

I think the above sample is enough I wrote that with `Vue.js` in [here](https://jsfiddle.net/qLy47hj2/1/). I am not sure the best way for Aurelia.

Please consider this feature. It is very important and useful for using HTML features besides Aurelia capability to create complete and powerful UI toolkit especially when you want to wrap some frameworks and libraries.

**Comment by fkleuver** [comment]
I like the idea of a special binding command for this behavior as @m-gallesio gave an example for. What do you think about this @HamedFathi, does it align with your original idea?

**Comment by HamedFathi** [comment]
@fkleuver 

Selective projection VS Complete projection

I believe an extended HTML input (as an Aurelia custom element) is an HTML input (with all attrs and events) so we should not select what we like or need, because the result is not a standard HTML input. This is a more complete version of the previous Aurelia custom element, not more. Of course, choosing is good in every situation. Is it possible to have both ideas together?

**Comment by fkleuver** [comment]
I must admit I have a little bit of difficulty precisely parsing your RFC (could be just my lack of concentration at the moment), would you mind showing me two html snippets representing a use case "before" and "after" (where both would have identical functionality) as you'd like to have it?

Just so there are no misunderstandings.
e.g.

before:
**app.html**
```html
<foo value.bind="value"></foo>
```
**foo.ts**
```html
export class FooCustomElement {
  @bindable value;
}
```
**foo.html**
```html
<template><input value.bind="value"></template>
```

after:
??

**Comment by HamedFathi** [comment]
@fkleuver 

Let me explain more.

First, the idea cames from Vue you can find it [here](https://vuejs.org/v2/guide/components-custom-events.html#Binding-Native-Events-to-Components)

with this

```html
<label>
  {{ label }}
  <input
    v-bind="$attrs" <<<<< HERE They project all native attributes of input for the ne Vue component.
    v-bind:value="value"
    v-on:input="$emit('input', $event.target.value)"
  >
</label>
```

Second, @bigopon implemented this idea [here](https://discourse.aurelia.io/t/assign-all-new-attributes-to-specific-element-transparent-wrapper-components/1949/9) in vCurrent with `au-bind` attribute.
About this 

1. It is better to have this in Aurelia vNext framework as an embedded feature.
2. With this implementation, we should use placeholder.bind="?" It is not cool because `placeholder` or `maxlength` or ... are HTML attributes, not Aurelia attributres so it is better to have them without using `.bind` or ...

Third, You want to know what is after it is so simple

Let investigate your sample 

```html
<foo value.bind="value"></foo>
```
What is `foo`? I suppose it is an extended HTML input like my [float-input](https://aurelia-toolbelt.github.io/#/bootstrap/float-input)

So this is just input with more options. Do you aggre with me? For user `foo` should work like `input` without lack of attributes.

BUT

the problem is here @fkleuver  Can you tell me based on your sample `foo` supports `placeholder` or `maxlength` attributes? Can you imagine an input element without them? Can you imagine a native HTML tag without the basic attributes? You should define all of them one by one in your new custom element Yes? Am I right?

So about after

I want to have this
```html
<foo value.bind="value" maxlength="10" placeholder='email@example.com'></foo>
```
Does it work now ? without attributes projection or define @bind one by one for each attributes I dont think so.

About [float-input](https://aurelia-toolbelt.github.io/#/bootstrap/float-input) in Aurelia-Toolbelt you can find the problem [float-input](https://aurelia-toolbelt.github.io/#/bootstrap/float-input)
It supports `placeholder` because I defined it but It does not support `maxlength` because I forgot to define it!

Does it make sense?

**Comment by fkleuver** [comment]
I think I get it, but I'm curious about various edge cases now. What needs to pass through and what doesn't.

Let's see if I understood it correctly. Going on with your last example. The current situation is this:

**foo.html**
All attribute must be bound to properties coming from the custom element VM
```html
<template>
  <input value.bind="value" maxlength.one-time="maxlength" placeholder.one-time="placeholder">
</template>
```
**foo.js**
All bindables need to be declared on the custom element VM
```ts
export class FooCustomElement {
  @bindable value;
  @bindable maxlength;
  @bindable placeholder;
}
```

**app.html**
And this is how you consume it
```html
<foo value.bind="value" maxlength="10" placeholder='email@example.com'></foo>
```

Now extrapolating on your proposal, how about this, is this also what you're looking for?

**foo.html**
No bindings declared on the input element whatsoever
```html
<template><input au-bind></template>
```
**foo.js**
No bindables declared on the custom element whatsoever (in fact it might as well be a html-only element)
```ts
export class FooCustomElement { }
```

**app.html**
And this is how you consume it (unchanged from the previous example, same effect)
```html
<foo value.bind="value" maxlength="10" placeholder='email@example.com'></foo>
```

In short, au-bind effectively says "take all bindings and attributes on the custom element and glue them to this element"

With the potential binding command variations for either opt-in or opt-out:
```html
<!-- only value and maxlength are bound -->
<template><input au-bind.only="value,maxlength"></template>
```

```html
<!-- everything except placeholder is bound -->
<template><input au-bind.except="placeholder"></template>
```

Effectively this makes it somewhat similar to `replaceable` but only the attributes are (replaced or added? what about conflicts) instead of the whole element. Is that what you'd like to have?

**Comment by HamedFathi** [comment]
@fkleuver 

> In short, au-bind effectively says "take all bindings and attributes on the custom element and glue them to this element"

Yes, take all **native HTML** bindings and attributes. For other attributes, we should use @bind like what we have now.

I realy like your idea about the selective mode 

```html
<!-- only value and maxlength are bound -->
<template><input au-bind.only="value,maxlength"></template>

and even this

<!-- everything except placeholder is bound -->
<template><input au-bind.except="placeholder"></template>

This is great because it covers all my needs and @m-gallesio too. You are awesome :)

**Comment by fkleuver** [comment]
> You are awesome :)

So are you. I really appreciate the discussions :)

> About conflicts, I think the priority is @bindable If I defined a binding attribute Aurelia should use that otherwise use au-bind attributes.

There wouldn't be a functional difference apart from the binding having 2 hops to go through instead of 1 (and the property being accessible + updatable from the view model), right?

Speaking of which, with `au-bind`, should those bindables be by convention also go through the view model or not? Perhaps another toggle or decorator to put on the VM? Just thinking out loud.

**Comment by HamedFathi** [comment]
@fkleuver 

> There wouldn't be a functional difference apart from the binding having 2 hops to go through instead of 1 (and the property being accessible + updatable from the view model), right?

Unfortunately, I did not get your point. May I ask you to explain more?

> Speaking of which, with au-bind, should those bindables be by convention also go through the view model or not? Perhaps another toggle or decorator to put on the VM? Just thinking out loud.

I think we should do everything about `au-bind` like `@bindable` variables. `au-bind` is just `less write do more`.

@shahabganji 

> The framework should throw an exception/warning saying: The property max-length might have conflicts at run time due to 'au-bind' and 'maxLength' bindable, in my opinion.

I am agree with you to throw exception when a developer handle a bindable variable twice. 
For example 
```ts
export class FooCustomElement{
 @bindable maxLength: number;
}
and
<template>
    <input au-bind.only="maxLength" >
</template>

The developer should add the `maxLength` to `au-bind.except`.

> I expect that placeholder and max-length being applied on the input and click be applied on the button, however the enabled attribute should be applied on both.

I think It should be possible because each element has own `au-bind` in your scenarios.
